File: assembly/calibrations/AngleOfCameraFrameInRefFrame_dA.cc
#include <iostream>
#include <cstdlib>
#include <cmath>

#include <TFile.h>
#include <TGraph.h>

/////////////////////////////////////////////////////////////////////////////////

// MEASUREMENTS
const double ANGDEG_PATREC = -0.275; //Average the best-match angles obtained via PatRec on the 2 fiducials <-> indicates the orientation of the sensor within the laboratory frame (?)

/*
const int PR1_PIXEL_X = 694;
const int PR1_PIXEL_Y = 867;

const double PR1_MOTION_STAGE_POS_X = -59.78;
const double PR1_MOTION_STAGE_POS_Y = -82.04;

const int PR2_PIXEL_X = 347;
const int PR2_PIXEL_Y = 534;

const double PR2_MOTION_STAGE_POS_X = 37.22;
const double PR2_MOTION_STAGE_POS_Y = -82.04;
*/

const int PR1_PIXEL_X = 1588;
const int PR1_PIXEL_Y = 605;

const double PR1_MOTION_STAGE_POS_X = -11.8108;
const double PR1_MOTION_STAGE_POS_Y = -116.634;

const int PR2_PIXEL_X = 1655;
const int PR2_PIXEL_Y = 685;

const double PR2_MOTION_STAGE_POS_X = 109.311;
const double PR2_MOTION_STAGE_POS_Y = -116.634;

// SENSOR, IMAGE SPECIFICATIONS - using markers in one line
const double ANGDEG_REFERC = 0.00; // PS-s //Target angle between 2 markers (?) <-> 0 by default, because the ref. points of both fiducials are supposed to be on the same line (both in laboratory/camera frames)
//const double ANGDEG_REFERC = 0.00; // PS-p

const double MM_PER_PIXEL = 0.0012;

const int IMAGE_PIXEL_WIDTH  = 2560;
const int IMAGE_PIXEL_HEIGHT = 1920;

// ANALYSIS PARAMETERS
const double ANGDEG_OFFSET_MIN  = -180.;
const double ANGDEG_OFFSET_MAX  = +180.;
const double ANGDEG_OFFSET_STEP = 0.005;

// OUTPUT OPTIONS
const std::string OUTPUT_TFILE_PATH = "AngleOfCameraFrameInRefFrame_dA.root";
const std::string TGRAPH_NAME       = "AngleOfCameraFrameInRefFrame_dA";

const int VERBOSE = 0;

/////////////////////////////////////////////////////////////////////////////////

class Point2D
{
 public:
  explicit Point2D(const double x, const double y): x_(x), y_(y) {}
  virtual ~Point2D() {}

  Point2D rotate_deg(const double) const;
  Point2D rotate_rad(const double) const;

  Point2D operator+(const Point2D&) const;

  void print() const;

  double x() const { return x_; }
  double y() const { return y_; }

 protected:
  double x_;
  double y_;
};

Point2D Point2D::rotate_deg(const double ang_deg) const
{
  const double ang_rad = ang_deg * (M_PI / 180.);

  return this->rotate_rad(ang_rad);
}

Point2D Point2D::rotate_rad(const double ang_rad) const
{
  const double x1 = (std::cos(ang_rad) * this->x()) - (std::sin(ang_rad) * this->y());
  const double y1 = (std::sin(ang_rad) * this->x()) + (std::cos(ang_rad) * this->y());

  return Point2D(x1, y1);
}

void Point2D::print() const
{
  std::cout << "(x=" << x_ << ", y=" << y_ << ")" << std::endl;

  return;
}

Point2D Point2D::operator+(const Point2D& p0) const
{
  return Point2D((this->x() + p0.x()), (this->y() + p0.y()));
}

Point2D from_pixels_to_distance_from_image_center(const int pixel_x, const int pixel_y)
{
  const double dx = double( pixel_x - (IMAGE_PIXEL_WIDTH  / 2.)) * MM_PER_PIXEL;
  const double dy = double(-pixel_y + (IMAGE_PIXEL_HEIGHT / 2.)) * MM_PER_PIXEL;

  return Point2D(dx, dy);
}

int main(int /* argc */, char** /* argv */)
{
  const Point2D d1 = from_pixels_to_distance_from_image_center(PR1_PIXEL_X, PR1_PIXEL_Y);
  const Point2D d2 = from_pixels_to_distance_from_image_center(PR2_PIXEL_X, PR2_PIXEL_Y);

  TFile ofile(OUTPUT_TFILE_PATH.c_str(), "recreate");

  TGraph g_offset;
  g_offset.SetName(TGRAPH_NAME.c_str());

  float best_angle = 999.;
  float best_angle_FOM = -1.0; // positive-def

  for(float angdeg_offset = ANGDEG_OFFSET_MIN; angdeg_offset <= ANGDEG_OFFSET_MAX; angdeg_offset += ANGDEG_OFFSET_STEP)
  {
    if(VERBOSE > 2){ std::cout << "* ANGLE_REFERC [deg] = " << ANGDEG_REFERC << std::endl; }
    if(VERBOSE > 2){ std::cout << "* ANGLE_PATREC [deg] = " << ANGDEG_PATREC << std::endl; }

    const Point2D d1_mm = d1.rotate_deg(angdeg_offset);
    const Point2D d2_mm = d2.rotate_deg(angdeg_offset);

    if(VERBOSE > 3){ std::cout << "* PatRec #1: center-to-bestmatch distance [mm] = "; d1_mm.print(); }
    if(VERBOSE > 3){ std::cout << "* PatRec #2: center-to-bestmatch distance [mm] = "; d2_mm.print(); }

    const Point2D PR1 = d1_mm + Point2D(PR1_MOTION_STAGE_POS_X, PR1_MOTION_STAGE_POS_Y);
    const Point2D PR2 = d2_mm + Point2D(PR2_MOTION_STAGE_POS_X, PR2_MOTION_STAGE_POS_Y);

    if(VERBOSE > 3){ std::cout << "* PatRec #1: best-match position [mm] = "; PR1.print(); }
    if(VERBOSE > 3){ std::cout << "* PatRec #2: best-match position [mm] = "; PR2.print(); }

    const double angdeg_measur = std::atan((PR2.y() - PR1.y()) / (PR2.x() - PR1.x())) * (180. / M_PI); //Compute angle between ref. points of both fiducials from their x-y positions (infered from motion stage position + best-match pixels in camera image) -- FOR CURRENTLY-TESTED ANGLE OFFSET 'angdeg_offset' !

    if(VERBOSE > 3){ std::cout << "* measured angle from PR1-best-match to PR2-best-match [deg] = " << angdeg_measur << std::endl; }

    const double angdeg_offset_new = angdeg_measur - ANGDEG_REFERC - ANGDEG_PATREC - 90.0; //If the currently-tested angle offset 'angdeg_offset' is correct, we expect the difference between 'ANGDEG_PATREC' (indicates sensor's rotation in laboratory frame) and 'angdeg_measur' (indicates sensor's rotation according to camera image) to match the expected 'ANGDEG_REFERC' value (=0) ! The 90degrees offset is due to the fact that the camera and motion stage axes are (historically ?) offset by 90 degrees, cf. camera images.

    if(VERBOSE > 2){ std::cout << "* camera-offset angle assumed [deg] = " << angdeg_offset     << std::endl; }
    if(VERBOSE > 2){ std::cout << "* camera-offset angle derived [deg] = " << angdeg_offset_new << std::endl; }

    const double angdeg_offset_diff = std::abs(angdeg_offset_new - angdeg_offset);

    if(VERBOSE > 2){ std::cout << "* camera-offset angle, abs-difference \"derived-assumed\" = " << angdeg_offset_diff << std::endl; }

    if(VERBOSE > 2){ std::cout << "-----------------------------------------------------------------------------------" << std::endl << std::endl; }

    if((angdeg_offset_diff < best_angle_FOM) || (best_angle_FOM < 0.))
    {
      best_angle     = angdeg_offset;
      best_angle_FOM = angdeg_offset_diff;
    }

    g_offset.SetPoint(g_offset.GetN(), angdeg_offset, angdeg_offset_diff);
  }

  if(VERBOSE > -1)
  {
    std::cout << "* camera frame angle in motion stage reference frame [deg] = " << best_angle << " (FOM=" << best_angle_FOM << ")" << std::endl;
  }

  ofile.cd();
  g_offset.Write();

  ofile.Close();

  return 0;
}

File: assembly/calibrations/AngleOfCameraFrameInRefFrame_dA.py
import math
import numpy as np
import matplotlib.pyplot as plt

# constants
mm_per_pixel = 0.0012
n_pixel_x = 2560
n_pixel_y = 1920


class Point2D:

    def __init__(self, xin, yin):
        self.x = xin
        self.y = yin

    def rotate_rad(self, angle_rad):
        x1 = (math.cos(angle_rad) * self.x) - (math.sin(angle_rad) * self.y)
        y1 = (math.sin(angle_rad) * self.x) + (math.cos(angle_rad) * self.y)

        return Point2D(x1, y1)

    def rotate_deg(self, angle_deg):
        return self.rotate_rad(angle_deg * math.pi / 180.)

    def __str__(self):
        return "(x=" + self.x + ", y=" + self.y + ")"

    def __add__(self, p0):
        return Point2D(self.x + p0.x, self.y + p0.y)


def from_pixels_to_distance_from_image_center(pixel_x, pixel_y):
    px = (pixel_x - (n_pixel_x / 2.)) * mm_per_pixel
    py = (-pixel_y + (n_pixel_y / 2.)) * mm_per_pixel

    return Point2D(px, py)


if __name__ == "__main__":

    pr1_ang_deg = float(
        input("Input best-match angle (in deg) for PatRec on TL marker: "))
    pr1_x_pixel = int(input(
        "Input x coordinate result (in pixels) of PatRec on TL marker (see printout on terminal): "))
    pr1_y_pixel = int(input(
        "Input y coordinate result (in pixels) of PatRec on TL marker (see printout on terminal): "))
    pr1_x_stage = float(input(
        "Input absolute x position (in mm) of motion stage for PatRec on TL marker: "))
    pr1_y_stage = float(input(
        "Input absolute y position (in mm) of motion stage for PatRec on TL marker: "))

    pr2_ang_deg = float(
        input("Input best-match angle (in deg) for PatRec on TR marker: "))
    pr2_x_pixel = int(input(
        "Input x coordinate result (in pixels) of PatRec on TR marker (see printout on terminal): "))
    pr2_y_pixel = int(input(
        "Input y coordinate result (in pixels) of PatRec on TR marker (see printout on terminal): "))
    pr2_x_stage = float(input(
        "Input absolute x position (in mm) of motion stage for PatRec on TR marker: "))
    pr2_y_stage = float(input(
        "Input absolute y position (in mm) of motion stage for PatRec on TR marker: "))

    pr_mean_ang_deg = (pr1_ang_deg + pr2_ang_deg) / 2.

    # Start calculations

    d1 = from_pixels_to_distance_from_image_center(pr1_x_pixel, pr1_y_pixel)
    d2 = from_pixels_to_distance_from_image_center(pr2_x_pixel, pr2_y_pixel)

    def calc_angle_offset_diff(angle_offset):
        d1_mm = d1.rotate_deg(angle_offset) + Point2D(pr1_x_stage, pr1_y_stage)
        d2_mm = d2.rotate_deg(angle_offset) + Point2D(pr2_x_stage, pr2_y_stage)

        ang_deg_meas = math.atan(
            (d2_mm.y - d1_mm.y) / (d2_mm.x - d1_mm.x)) * 180. / math.pi

        ang_deg_offset_new = ang_deg_meas - pr_mean_ang_deg - 90.

        ang_deg_offset_diff = abs(ang_deg_offset_new - angle_offset)

        return ang_deg_offset_diff

    offset_arr = np.arange(-180, 180, 0.005)
    angle_diff_arr = np.zeros(offset_arr.size)

    for i, offset in enumerate(offset_arr):
        angle_diff_arr[i] = calc_angle_offset_diff(offset)

    best_angle_FOM = np.min(angle_diff_arr)
    best_angle_idx = np.argmin(angle_diff_arr)
    best_angle = offset_arr[best_angle_idx]

    print("  Camera frame angle in motion stage ref frame [deg] = " + str(
        best_angle) + "  (FOM= " + str(best_angle_FOM) + ")")

    # Plotting result

    fig, axs = plt.subplots(1, 2, figsize=(12, 6))
    axs[0].plot(offset_arr, angle_diff_arr, '.c')
    axs[0].set_xlabel("Offset angle [deg]")
    axs[0].set_ylabel("Computed difference from target [deg]")

    axs[1].plot(offset_arr, angle_diff_arr, '.c')
    axs[1].set(xlim=(best_angle-0.1, best_angle+0.1), ylim=(best_angle_FOM,
               np.max([angle_diff_arr[best_angle_idx-20], angle_diff_arr[best_angle_idx+20]])))
    axs[1].set_xlabel("Offset angle [deg]")
    axs[1].set_ylabel("Computed difference from target [deg]")

    plt.ion()
    plt.show()

    input("\nPress enter to close")
